src/rabbit_vhost.erl

```diff
@@ -213,92 +229,95 @@ assert_benign({error, {absent, Q, _}}, ActingUser) ->
+
+exists(VHost) ->
+    mnesia:dirty_read({rabbit_vhost, VHost}) /= [].
 
-exists(VHostPath) ->
-    mnesia:dirty_read({rabbit_vhost, VHostPath}) /= [].
+-spec list_names() -> [vhost:name()].
 
--spec list() -> [rabbit_types:vhost()].
+list_names() -> mnesia:dirty_all_keys(rabbit_vhost).
 
-list() ->
-    mnesia:dirty_all_keys(rabbit_vhost).
+-spec list() -> [vhost:name()].
+
+list() -> mnesia:dirty_match_object(rabbit_vhost, vhost:pattern_match_all()).
 
 -spec count() -> non_neg_integer().
 count() ->
     length(list()).
 
--spec with(rabbit_types:vhost(), rabbit_misc:thunk(A)) -> A.
+-spec with(vhost:name(), rabbit_misc:thunk(A)) -> A.
 
-with(VHostPath, Thunk) ->
+with(VHost, Thunk) ->
     fun () ->
-            case mnesia:read({rabbit_vhost, VHostPath}) of
+            case mnesia:read({rabbit_vhost, VHost}) of
                 [] ->
-                    mnesia:abort({no_such_vhost, VHostPath});
+                    mnesia:abort({no_such_vhost, VHost});
                 [_V] ->
                     Thunk()
             end
     end.
 
 -spec with_user_and_vhost
-        (rabbit_types:username(), rabbit_types:vhost(), rabbit_misc:thunk(A)) -> A.
+        (rabbit_types:username(), vhost:name(), rabbit_misc:thunk(A)) -> A.
 
-with_user_and_vhost(Username, VHostPath, Thunk) ->
-    rabbit_misc:with_user(Username, with(VHostPath, Thunk)).
+with_user_and_vhost(Username, VHost, Thunk) ->
+    rabbit_misc:with_user(Username, with(VHost, Thunk)).
 
 %% Like with/2 but outside an Mnesia tx
 
--spec assert(rabbit_types:vhost()) -> 'ok'.
+-spec assert(vhost:name()) -> 'ok'.
 
-assert(VHostPath) -> case exists(VHostPath) of
+assert(VHost) -> case exists(VHost) of
                          true  -> ok;
-                         false -> throw({error, {no_such_vhost, VHostPath}})
+                         false -> throw({error, {no_such_vhost, VHost}})
                      end.
 
--spec update(rabbit_types:vhost(), fun((#vhost{}) -> #vhost{})) -> #vhost{}.
+-spec update(vhost:name(), fun((vhost:vhost()) -> vhost:vhost())) -> vhost:vhost().
 
-update(VHostPath, Fun) ->
-    case mnesia:read({rabbit_vhost, VHostPath}) of
+update(VHost, Fun) ->
+    case mnesia:read({rabbit_vhost, VHost}) of
         [] ->
-            mnesia:abort({no_such_vhost, VHostPath});
+            mnesia:abort({no_such_vhost, VHost});
         [V] ->
             V1 = Fun(V),
             ok = mnesia:write(rabbit_vhost, V1, write),
             V1
     end.
 
-limits_of(VHostPath) when is_binary(VHostPath) ->
-    assert(VHostPath),
-    case mnesia:dirty_read({rabbit_vhost, VHostPath}) of
+limits_of(VHost) when is_binary(VHost) ->
+    assert(VHost),
+    case mnesia:dirty_read({rabbit_vhost, VHost}) of
         [] ->
-            mnesia:abort({no_such_vhost, VHostPath});
-        [#vhost{limits = Limits}] ->
-            Limits
+            mnesia:abort({no_such_vhost, VHost});
+        [VHost] ->
+            ?vhost_field_limits(VHost)
```

Comment:
As said earlier in this review, please use a function call to the `vhost` module, not inlined code.

This comment is valid for all the uses of those macros in function bodies.